smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/SymbolVisitor.java

```diff
@@ -378,18 +396,63 @@ private String flattenShapeName(ToShapeId id) {
                 .definitionFile(toFilename(namespace));
     }
 
-    private String formatModuleName(ShapeType shapeType, String name) {
-        // All shapes except for the service and operations are stored in models.
-        if (shapeType == ShapeType.SERVICE) {
-            return "./" + name;
-        } else if (shapeType == ShapeType.OPERATION) {
-            return "./commands/" + name;
-        } else {
-            return "./models/index";
-        }
-    }
-
     private String toFilename(String namespace) {
         return namespace + ".ts";
     }
+
+    /**
+     * Utility class to locate which path should the symbol be generated into.
+     * It will break the models into multiple files to prevent it being too big.
+     */
+    static final class ModuleNameDelegator {
+        static final int DEFAULT_CHUNK_SIZE = 300;
+        static final String SHAPE_NAMESPACE_PREFIX = "./models/";
+
+        private final Map<Shape, String> visitedModels = new HashMap<>();
+        private int bucketCount = 0;
+        private int currentBucketSize = 0;
+        private final int chunkSize;
+
+        ModuleNameDelegator(int shapeChunkSize) {
+            chunkSize = shapeChunkSize;
+        }
+
+        ModuleNameDelegator() {
```

Comment:
Note sure you even need a defualt constructor here since this class isn't exposed publicly